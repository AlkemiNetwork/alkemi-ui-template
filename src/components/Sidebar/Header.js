import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import IconButton from "@material-ui/core/IconButton";
import ChevronLeftIcon from "@material-ui/icons/ChevronLeft";

const styles = theme => ({
    drawerHeader: {
        display: "flex",
        alignItems: "center",
        justifyContent: "flex-end",
        padding: "0 " + theme.spacing.unit + "px",
        ...theme.mixins.toolbar
    }
});

const SidebarHeader = ({ classes, onClick }) => {
    return (
        <div className={classes.drawerHeader}>
            <IconButton onClick={onClick}>
                <ChevronLeftIcon />
            </IconButton>
        </div>
    );
};

SidebarHeader.propTypes = {
    classes: PropTypes.object.isRequired,
    selected: PropTypes.bool,
    onClick: PropTypes.func.isRequired
};

export default withStyles(styles)(SidebarHeader);
