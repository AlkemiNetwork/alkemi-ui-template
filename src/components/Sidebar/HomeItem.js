import React from "react";
import PropTypes from "prop-types";

import HomeIcon from "@material-ui/icons/Home";

import ListItem1 from "../UI/ListItem/1";

const SidebarHomeItem = ({ onClick, selected }) => (
    <ListItem1
        icon={<HomeIcon />}
        text="Home"
        onClick={onClick}
        selected={selected}
    />
);

SidebarHomeItem.propTypes = {
    selected: PropTypes.bool,
    onClick: PropTypes.func.isRequired
};

export default SidebarHomeItem;
