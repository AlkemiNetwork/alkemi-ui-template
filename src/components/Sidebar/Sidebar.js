import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { compose } from "redux";

import Drawer from "@material-ui/core/Drawer";
import List from "@material-ui/core/List";
import withWidth, { isWidthUp } from "@material-ui/core/withWidth";
import { withStyles } from "@material-ui/core/styles";

import SidebarHomeItem from "../../containers/Sidebar/HomeItem";
import SidebarHeader from "../../containers/Sidebar/Header";

const styles = theme => ({
    drawer: {
        height: "100vh",
        position: "fixed"
    },
    drawerPaper: {
        position: "relative",
        height: "100%",
        width: 240,
        transition: theme.transitions.create("width", {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.enteringScreen
        })
    },
    drawerPaperClose: {
        width: 0,
        overflowX: "hidden",
        transition: theme.transitions.create("width", {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.leavingScreen
        }),
        [theme.breakpoints.up("sm")]: {
            width: 60
        }
    },
    drawerInner: {
        // Make the items inside not wrap when transitioning:
        width: 240
    }
});

const Sidebar = ({ classes, width, variant, open, onClose }) => {
    const isLargeScreen = isWidthUp("sm", width);
    return (
        <Drawer
            variant={isLargeScreen ? "permanent" : undefined}
            classes={
                isLargeScreen
                    ? {
                          docked: classes.drawer,
                          paper: classNames(
                              classes.drawerPaper,
                              !open && classes.drawerPaperClose
                          )
                      }
                    : {}
            }
            open={open}
            onClose={onClose}
        >
            <div className={classes.drawerInner}>
                <SidebarHeader />
                <List className={classes.list}>
                    <SidebarHomeItem />
                </List>
            </div>
        </Drawer>
    );
};

Sidebar.propTypes = {
    classes: PropTypes.object.isRequired,
    width: PropTypes.oneOf(["xs", "sm", "md", "lg", "xl"]),
    variant: PropTypes.string,
    open: PropTypes.bool,
    onClose: PropTypes.func
};

export default compose(withWidth(), withStyles(styles))(Sidebar);
