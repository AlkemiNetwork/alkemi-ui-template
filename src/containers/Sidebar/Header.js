import { connect } from "react-redux";

import { closeSidebar } from "../../redux/actions/ui";

import SidebarHeader from "../../components/Sidebar/Header";

const mapDispatchToProps = {
    onClick: closeSidebar
};

export default connect(undefined, mapDispatchToProps)(SidebarHeader);
