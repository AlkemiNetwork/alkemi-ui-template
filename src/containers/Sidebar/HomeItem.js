// Libs
import { connect } from "react-redux";
import { push } from "connected-react-router";
// Component
import SidebarHomeItem from "../../components/Sidebar/HomeItem";
// Constants
import { HOME } from "../../constants/routes";
// Selectors
import { isItemSelected } from "../../redux/selectors/ui";

const mapStateToProps = state => ({
    selected: isItemSelected(state)
});

const mapDispatchToProps = {
    onClick: () => push(HOME)
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(SidebarHomeItem);
