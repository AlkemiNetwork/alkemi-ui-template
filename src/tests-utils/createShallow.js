import { shallow as _shallow } from "enzyme";

const createShallow = ({ dive }) => {
    const shallow = (node, options) => {
        let wrapper = _shallow(node, options);

        for (let i = 0; i < dive; i++) {
            if (wrapper.name().match(/^WithWidth(.+)$/)) {
                // Diving through a Meterial-UI WidthWith component can not be performed
                // by default because of React.Fragment used in withWidth
                // It is a known issue: https://github.com/airbnb/enzyme/issues/1213
                wrapper = wrapper
                    .dive()
                    .children()
                    .at(0);
            } else {
                wrapper = wrapper.dive();
            }
        }
        return wrapper;
    };

    return shallow;
};

export default createShallow;
