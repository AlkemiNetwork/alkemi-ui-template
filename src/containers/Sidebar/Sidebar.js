import { connect } from "react-redux";

import { closeSidebar } from "../../redux/actions/ui";

import Sidebar from "../../components/Sidebar/Sidebar";

const mapStateToProps = (state, ownProps) => ({
    open: state.ui.sidebar.open
});

const mapDispatchToProps = {
    onClose: closeSidebar
};

export default connect(mapStateToProps, mapDispatchToProps)(Sidebar);
