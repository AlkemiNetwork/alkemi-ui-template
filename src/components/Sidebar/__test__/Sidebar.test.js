import React from "react";
import toJson from "enzyme-to-json";

import Drawer from "@material-ui/core/Drawer";
import { MuiThemeProvider, createMuiTheme } from "@material-ui/core/styles";

import Sidebar from "../Sidebar";
import SidebarHomeItem from "../../../containers/Sidebar/HomeItem";
import SidebarHeader from "../../../containers/Sidebar/Header";
import createShallow from "../../../tests-utils/createShallow";

describe("<Sidebar />", () => {
    let shallow;
    let wrapper;
    let props = {};

    beforeEach(() => {
        props = {
            onClose: jest.fn()
        };
        shallow = createShallow({ dive: 3 });
        wrapper = shallow(<Sidebar {...props} />);
    });

    it("renders 1 <Drawer /> components", () => {
        expect(wrapper.find(Drawer).length).toEqual(1);
    });

    it("renders 1 <SidebarHeader /> components", () => {
        expect(wrapper.find(SidebarHeader).length).toEqual(1);
    });

    it("renders 1 <SidebarHomeItem /> components", () => {
        expect(wrapper.find(SidebarHomeItem).length).toEqual(1);
    });

    it("button is functional", () => {
        wrapper
            .find(Drawer)
            .at(0)
            .simulate("close");
        expect(props.onClose.mock.calls.length).toEqual(1);
    });

    describe("when width is smUp", () => {
        beforeEach(() => {
            global.window.innerWidth = 600;
            shallow = createShallow({ dive: 4 });
            wrapper = shallow(
                <MuiThemeProvider
                    theme={createMuiTheme({
                        breakpoints: { values: { sm: 600 } }
                    })}
                >
                    <Sidebar {...props} />
                </MuiThemeProvider>
            );
        });

        it("matches snapshot", () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it("Drawer has class drawer and paper", () => {
            expect(
                wrapper
                    .find(Drawer)
                    .at(0)
                    .props().classes.docked
            ).toBeDefined();
            expect(
                wrapper
                    .find(Drawer)
                    .at(0)
                    .props().classes.paper
            ).toBeDefined();
        });
    });

    describe("when width is smDown", () => {
        beforeEach(() => {
            global.window.innerWidth = 599;
            shallow = createShallow({ dive: 4 });
            wrapper = shallow(
                <MuiThemeProvider
                    theme={createMuiTheme({
                        breakpoints: { vaues: { sm: 600 } }
                    })}
                >
                    <Sidebar {...props} />
                </MuiThemeProvider>
            );
        });

        it("matches snapshot", () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });

        it("Drawer has not class drawer nor paper", () => {
            expect(
                wrapper
                    .find(Drawer)
                    .at(0)
                    .props().classes.docked
            ).toBeFalsy();
            expect(
                wrapper
                    .find(Drawer)
                    .at(0)
                    .props().classes.paper
            ).toBeFalsy();
        });
    });
});
