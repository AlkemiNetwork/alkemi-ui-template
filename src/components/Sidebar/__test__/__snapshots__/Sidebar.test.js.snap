// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> when width is smDown matches snapshot 1`] = `
<WithStyles(Drawer)
  classes={Object {}}
  onClose={[MockFunction]}
>
  <div
    className="Sidebar-drawerInner-4"
  >
    <Connect(WithStyles(SidebarHeader)) />
    <WithStyles(List)>
      <Connect(SidebarHomeItem) />
    </WithStyles(List)>
  </div>
</WithStyles(Drawer)>
`;

exports[`<Sidebar /> when width is smUp matches snapshot 1`] = `
<WithStyles(Drawer)
  classes={
    Object {
      "docked": "Sidebar-drawer-1",
      "paper": "Sidebar-drawerPaper-2 Sidebar-drawerPaperClose-3",
    }
  }
  onClose={[MockFunction]}
  variant="permanent"
>
  <div
    className="Sidebar-drawerInner-4"
  >
    <Connect(WithStyles(SidebarHeader)) />
    <WithStyles(List)>
      <Connect(SidebarHomeItem) />
    </WithStyles(List)>
  </div>
</WithStyles(Drawer)>
`;
